include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +393,277 @@ class BidirectionalNode
+  NewNode(const NewNode &) = default;
+  NewNode(NewNode &&) = default;
+  NewNode &operator=(const NewNode &) = default;
+  NewNode &operator=(NewNode &&) = default;
+
+public:
+  // This stuff is needed by the DominatorTree implementation
+  void printAsOperand(llvm::raw_ostream &, bool) const { revng_abort(); }
+
+public:
+  EdgeView addSuccessor(NewNode &NewSuccessor, EdgeLabel EL = EdgeLabel{}) {
+    OwningEdge Owner{ &NewSuccessor,
+                      std::make_unique<EdgeLabel>(std::move(EL)) };
+    NonOwningEdge View{ this, Owner.Label.get() };
+    auto &Output = Successors.emplace_back(std::move(Owner));
+    NewSuccessor.Predecessors.emplace_back(std::move(View));
+    return Output;
+  }
+  EdgeView addPredecessor(NewNode &NewPredecessor, EdgeLabel EL = EdgeLabel{}) {
+    OwningEdge Owner{ this, std::make_unique<EdgeLabel>(std::move(EL)) };
+    NonOwningEdge View{ &NewPredecessor, Owner.Label.get() };
+    auto &Output = NewPredecessor.Successors.emplace_back(std::move(Owner));
+    Predecessors.emplace_back(std::move(View));
+    return Output;
+  }
+
+public:
+  auto successor_edges() {
+    auto ToView = [](auto &E) -> EdgeView { return E; };
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return llvm::map_range(Range, ToView);
+  }
+  auto successor_edges() const {
+    auto ToView = [](auto const &E) -> EdgeView const { return E; };
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return llvm::map_range(Range, ToView);
+  }
+  auto predecessor_edges() {
+    auto ToView = [](auto &E) -> EdgeView { return E; };
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return llvm::map_range(Range, ToView);
+  }
+  auto predecessor_edges() const {
+    auto ToView = [](auto const &E) -> EdgeView const { return E; };
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return llvm::map_range(Range, ToView);
+  }
+
+  auto successors() {
+    auto ToNeighbor = [](auto &E) -> auto * { return E.Neighbor; };
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return llvm::map_range(Range, ToNeighbor);
+  }
+  auto successors() const {
+    auto ToNeighbor = [](auto const &E) -> auto const * { return E.Neighbor; };
+    auto Range = llvm::make_range(Successors.begin(), Successors.end());
+    return llvm::map_range(Range, ToNeighbor);
+  }
+  auto predecessors() {
+    auto ToNeighbor = [](auto &E) -> auto * { return E.Neighbor; };
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return llvm::map_range(Range, ToNeighbor);
+  }
+  auto predecessors() const {
+    auto ToNeighbor = [](auto const &E) -> auto const * { return E.Neighbor; };
+    auto Range = llvm::make_range(Predecessors.begin(), Predecessors.end());
+    return llvm::map_range(Range, ToNeighbor);
+  }
+
+protected:
+  typename EdgeOwnerContainer::iterator findSuccessor(DerivedType const &S) {
```

Comment:
`auto`?

Reply:
`auto` is a good idea, I just thought since it was a protected function, it could use a bit of extra explicitness, but yeah, `auto` is better in most cases.